### PR TITLE
Go SDK avroio Package - Read/Write Avro Files 

### PR DESCRIPTION
**Pull request by daidokoro**
Implemented an io package for reading/writing __avro__ files. This implementation borrows directly from the textio package, allowing it to support all the various read/write methodologies utilised in that package. I.e local storage, GCS, etc

_usage:_
```Go
package main

import (
	"context"
	"flag"
	"log"
	"reflect"

	"github.com/apache/beam/sdks/go/pkg/beam"
	"github.com/apache/beam/sdks/go/pkg/beam/x/beamx"
	"github.com/daidokoro/beam/sdks/go/pkg/beam/io/avroio"
	"github.com/daidokoro/beam/sdks/go/pkg/beam/x/debug"
)

// Simple Avro Read Example - sample avro file provided

// Doc type used to unmarshal avro json data

type Doc struct {
	Stamp int64  `json:"timestamp"`
	Tweet string `json:"tweet"`
	User  string `json:"username"`
}

const schema = `{
	"type": "record",
	"name": "tweet",
	"namespace": "twitter",
	"fields": [
		{ "name": "timestamp", "type": "int" },
		{ "name": "tweet", "type": "string" },
		{ "name": "username", "type": "string" }
	]
}`

func main() {
	flag.Parse()
	beam.Init()

	ctx := context.Background()
	p := beam.NewPipeline()
	s := p.Root()

	rows := avroio.Read(s, "./twitter.avro", reflect.TypeOf(""))
	debug.Print(s, rows)

	docs := avroio.Read(s, "./twitter.avro", reflect.TypeOf(Doc{}))
	debug.Print(s, docs)

	format := beam.ParDo(s, func(d Doc, emit func(string)) {
		d.User = "daidokoro"
		d.Tweet = "I was here......"

		b, _ := json.Marshal(d)
		emit(string(b))
	}, docs)

	debug.Print(s, format)
	avroio.Write(s, "./out.avro", schema, format)

	if err := beamx.Run(context.Background(), p); err != nil {
		log.Fatalf(ctx, "Failed to execute job: %v", err)
	}
}

```

Example uses sample _twitter.avro_ file found [here](https://github.com/miguno/avro-cli-examples/blob/master/twitter.avro?raw=true)

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by daidokoro** [comment]
@lostluck please review? 
Thanks :-)

File: sdks/go/pkg/beam/io/avroio/avroio.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
+package avroio
```
[the remark this answers is not in the document]
Comment (daidokoro) [reply sdks/go/pkg/beam/io/avroio/avroio.go]: Thanks for the review.. Fixed this in 599cba62c381093feb6419782984b485b12f0f60

File: sdks/go/pkg/beam/io/avroio/avroio.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+
+// Avroio package based directly on textio with some
+// elements borrowed from the bigquery implementation.
+// Supports reading/writing and unmarshalling Avro files.
+// --
+
+func init() {
+	// beam.RegisterType(reflect.TypeOf((*writeFileFn)(nil)).Elem())
+	// beam.RegisterFunction(readFn)
+	beam.RegisterFunction(expandFn)
```
[the remark this answers is not in the document]
Comment (daidokoro) [reply sdks/go/pkg/beam/io/avroio/avroio.go]: Thanks for the info. I guess I left it in as this is how the textio package is defined.

File: sdks/go/pkg/beam/io/avroio/avroio.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (daidokoro) [reply sdks/go/pkg/beam/io/avroio/avroio.go]: I've added a Package Comment - 599cba62c381093feb6419782984b485b12f0f60
I've also added the example code to the go sdk examples sections, wasn't sure how to handle the sample twitter.avro file, so I've included a download link.

Let me know if this works

Thanks

**Comment by lostluck** [comment]
@herohde Please merge :)

**Pull request merged**